coreax/coresubset.py

```diff
@@ -617,3 +620,294 @@ def _loop_body(
+    :param score_method: :class:`~coreax.score_matching.ScoreMatching` object to use, or
+        :data:`None` (default), which will use kernel density matching
+    :param regularise: Boolean that enforces regularisation, as in
+        :cite:`benard2023kernel`.
+    """
+
+    def __init__(
+        self,
+        random_key: coreax.util.KeyArrayLike,
+        *,
+        kernel: coreax.kernel.Kernel,
+        weights_optimiser: coreax.weights.WeightsOptimiser | None = None,
+        block_size: int = 10_000,
+        unique: bool = True,
+        refine_method: coreax.refine.Refine | None = None,
+        score_method: coreax.score_matching.ScoreMatching | None = None,
+        regularise: bool = True,
+    ):
+        """Initialise a KernelHerding class."""
+        # Assign herding-specific attributes
+        self.block_size = block_size
+        self.unique = unique
+        self.random_key = random_key
+        self.regularise = regularise
+        self.score_method = score_method
+        self.stein_kernel = None
+
+        # Initialise parent
+        super().__init__(
+            weights_optimiser=weights_optimiser,
+            kernel=kernel,
+            refine_method=refine_method,
+        )
+
+    def tree_flatten(self) -> tuple[tuple, dict]:
+        """
+        Flatten a pytree.
+
+        Define arrays & dynamic values (children) and auxiliary data (static values).
+        A method to flatten the pytree needs to be specified to enable JIT decoration
+        of methods inside this class.
+
+        :return: Tuple containing two elements. The first is a tuple holding the arrays
+            and dynamic values that are present in the class. The second is a dictionary
+            holding the static auxiliary data for the class, with keys being the names
+            of class attributes, and values being the values of the corresponding class
+            attributes.
+        """
+        children = (
+            self.random_key,
+            self.kernel,
+            self.stein_kernel,
+            self.coreset_indices,
+            self.coreset,
+        )
+        aux_data = {
+            "block_size": self.block_size,
+            "unique": self.unique,
+            "refine_method": self.refine_method,
+            "weights_optimiser": self.weights_optimiser,
+            "regularise": self.regularise,
+            "score_method": self.score_method,
+        }
+        return children, aux_data
+
+    def fit_to_size(self, coreset_size: int) -> None:
+        r"""
+        Execute Stein thinning algorithm with Jax.
+
+        We first compute a score function, and then the Stein kernel. This is used to
+        greedily choose points in the coreset to minimise kernel Stein discrepancy
+        (KSD).
+
+        :param coreset_size: The size of the of coreset to generate
+        """
+        if coreset_size == 0:
+            raise ValueError("coreset_size must be non-zero")
+
```

Comment:
Same comment as in https://github.com/gchq/coreax/pull/572#discussion_r1579007131

Reply:
I've kept this one as the regularised version of Stein thinning needs to check for division by zero when setting the $\lambda$ parameter. The exception doesn't fire as is due to `jnp.zeros` accepting `0` as an argument in the `try; catch`.